jeff65/gold/grammar/Gold.py

```diff
@@ -0,0 +1,2507 @@
+        buf.write("\u00c9\u00c7\3\2\2\2\u00ca\u00cb\7+\2\2\u00cb\u00cd\5")
+        buf.write("\"\22\2\u00cc\u00ca\3\2\2\2\u00cc\u00cd\3\2\2\2\u00cd")
+        buf.write("\u00ce\3\2\2\2\u00ce\u00cf\7-\2\2\u00cf\31\3\2\2\2\u00d0")
+        buf.write("\u00d1\7#\2\2\u00d1\u00d2\7\3\2\2\u00d2\u00d3\5\"\22\2")
+        buf.write("\u00d3\u00d4\7/\2\2\u00d4\33\3\2\2\2\u00d5\u00d6\7!\2")
+        buf.write("\2\u00d6\u00d7\7\3\2\2\u00d7\u00e0\7\67\2\2\u00d8\u00dd")
+        buf.write("\5\f\7\2\u00d9\u00da\7\65\2\2\u00da\u00dc\5\f\7\2\u00db")
+        buf.write("\u00d9\3\2\2\2\u00dc\u00df\3\2\2\2\u00dd\u00db\3\2\2\2")
+        buf.write("\u00dd\u00de\3\2\2\2\u00de\u00e1\3\2\2\2\u00df\u00dd\3")
+        buf.write("\2\2\2\u00e0\u00d8\3\2\2\2\u00e0\u00e1\3\2\2\2\u00e1\u00e2")
+        buf.write("\3\2\2\2\u00e2\u00e5\78\2\2\u00e3\u00e4\7\66\2\2\u00e4")
+        buf.write("\u00e6\5\b\5\2\u00e5\u00e3\3\2\2\2\u00e5\u00e6\3\2\2\2")
+        buf.write("\u00e6\u00e7\3\2\2\2\u00e7\u00e8\5\"\22\2\u00e8\u00e9")
+        buf.write("\7.\2\2\u00e9\35\3\2\2\2\u00ea\u00ec\7%\2\2\u00eb\u00ed")
+        buf.write("\5\2\2\2\u00ec\u00eb\3\2\2\2\u00ec\u00ed\3\2\2\2\u00ed")
+        buf.write("\37\3\2\2\2\u00ee\u00ef\5\2\2\2\u00ef\u00f0\7\31\2\2\u00f0")
+        buf.write("\u00f1\5\2\2\2\u00f1\u00fb\3\2\2\2\u00f2\u00f3\5\2\2\2")
+        buf.write("\u00f3\u00f4\7\32\2\2\u00f4\u00f5\5\2\2\2\u00f5\u00fb")
+        buf.write("\3\2\2\2\u00f6\u00f7\5\2\2\2\u00f7\u00f8\7\33\2\2\u00f8")
+        buf.write("\u00f9\5\2\2\2\u00f9\u00fb\3\2\2\2\u00fa\u00ee\3\2\2\2")
+        buf.write("\u00fa\u00f2\3\2\2\2\u00fa\u00f6\3\2\2\2\u00fb!\3\2\2")
+        buf.write("\2\u00fc\u0105\5\16\b\2\u00fd\u0105\5\26\f\2\u00fe\u0105")
+        buf.write("\5\30\r\2\u00ff\u0105\5\22\n\2\u0100\u0105\5\36\20\2\u0101")
+        buf.write("\u0105\5\24\13\2\u0102\u0105\5 \21\2\u0103\u0105\5\2\2")
+        buf.write("\2\u0104\u00fc\3\2\2\2\u0104\u00fd\3\2\2\2\u0104\u00fe")
+        buf.write("\3\2\2\2\u0104\u00ff\3\2\2\2\u0104\u0100\3\2\2\2\u0104")
+        buf.write("\u0101\3\2\2\2\u0104\u0102\3\2\2\2\u0104\u0103\3\2\2\2")
+        buf.write("\u0105\u0108\3\2\2\2\u0106\u0104\3\2\2\2\u0106\u0107\3")
+        buf.write("\2\2\2\u0107#\3\2\2\2\u0108\u0106\3\2\2\2\u0109\u010f")
+        buf.write("\5\16\b\2\u010a\u010f\5\32\16\2\u010b\u010f\5\22\n\2\u010c")
+        buf.write("\u010f\5\20\t\2\u010d\u010f\5\34\17\2\u010e\u0109\3\2")
+        buf.write("\2\2\u010e\u010a\3\2\2\2\u010e\u010b\3\2\2\2\u010e\u010c")
+        buf.write("\3\2\2\2\u010e\u010d\3\2\2\2\u010f\u0112\3\2\2\2\u0110")
+        buf.write("\u010e\3\2\2\2\u0110\u0111\3\2\2\2\u0111\u0113\3\2\2\2")
+        buf.write("\u0112\u0110\3\2\2\2\u0113\u0114\7\2\2\3\u0114%\3\2\2")
+        buf.write("\2\35\61Z]`bknx\177\u0084\u008b\u009a\u00a1\u00aa\u00ac")
+        buf.write("\u00b9\u00c7\u00cc\u00dd\u00e0\u00e5\u00ec\u00fa\u0104")
+        buf.write("\u0106\u010e\u0110")
+        return buf.getvalue()
```

Comment:
It's the state machine for the parser in binary form, I think. That's a generated file, so it can neither be commented nor broken up, since changing the source file will just cause it to be overwritten.

I'm not generally a huge fan of checking in generated files, but in this case it seemed more convenient to do so.

Reply:
what are the steps to re-generate it?

it looks like it's ultimately string data as far as python is concerned, so even if we need to have it checked in, we should still be able to move its contents into a datafile and have this function read in the datafile and return its contents as a StringIO object.

oh, wait... unless this entire `Gold.py` file is generated? if so, what's your qualm with making generation of this (and other) files part of the build process and `.gitignore`ing the generated files? i don't really see an advantage to checking it in.

Reply:
Yeah, literally every file in the grammar subdirectory is generated except for `Gold.g4` and `__init__.py`. So the entirety of `Gold.py` is generated, for example.

I checked them in because in order to generate them, you need to have antlr4 installed. Not only that, but you need to have the version of antlr4 that matches the version of the library we use. Even worse, antlr4 is a Java application, so you need Java... and then of course setting all that up on Windows is a pain, and eventually the Gold.g4 file will be stable enough that it won't need to be regenerated very often.

On the other hand, the generated files are text files, so it's not like git is going to throw a fit from having them around; they'll diff efficiently.

Basically, there are no downsides to having them checked in other than them being a little ugly. The upside is that if you don't have to touch the grammar file, everything you need is already present in the repository.

Reply:
The steps to regenerate it is basically "get everything installed and run `make` in the checkout directory."

I wouldn't be opposed to replacing Antlr4 with some pure-python generator, or even Lark if we can convince it to play nice. Once the initial AST is constructed, there's no dependency on Antlr4, so replacing it doesn't require rewriting the entire compiler.

Reply:
okay, i'm fine with having the generated files in source control. i do, however, want to say that checking in generated files *does* have a downside: developers making changes directly to the generated file, because it seems like a natural thing to do. or the .g4 file and the generated file becoming out of sync. so i guess we just need to watch out for that.

you say the antlr version must be extremely specific... that raises some red flags for me. i've worked on a project that depended on a very specific version of a tool (except they didn't realize it was so version dependent) and the vendor only offered downloads for the latest version. nobody kept a copy of the installer for the version they were using. it was an absolute disaster, the build scripts had to be rewritten, and the entire team upgraded to the latest version of the tool and they stored a copy of the installer in a safe place. obviously that wasn't the best solution, but they were already too dependent on the tool.
you say the ast is generic enough that our dependency on antlr is relatively loose, so that's encouraging. lets just try to keep it that way.